src/lib/rootfs/image/image.c

```diff
@@ -79,9 +84,64 @@ int rootfs_image_init(char *source, char *mount_dir) {
         }
         read_write = 1;
     } else {
-        if ( ( image_fp = fopen(source, "r") ) == NULL ) { // Flawfinder: ignore
-            singularity_message(ERROR, "Could not open image (read only) %s: %s\n", source, strerror(errno));
-            ABORT(255);
+        if ( ( image_fp = fopen(source, "r") ) != NULL ) { // Flawfinder: ignore
+            singularity_message(VERBOSE, "Opened image (read only, without privileges) %s\n", source);
+        } else {
+            singularity_config_rewind();
+            if ( singularity_config_get_bool("allow privileged fopen", 0) == 0 ) {
+                singularity_message(ERROR, "Privileged fopen is not enabled in config\n" );
+                ABORT(255);
+            }
+
+            singularity_priv_escalate();
+            if ( ( image_fp = fopen(source, "r") ) == NULL ) { // Flawfinder: ignore
+	            singularity_message(ERROR, "Could not open image (read only, with privileges) %s: %s\n", source, strerror(errno));
+                ABORT(255);
+            }
+
+            singularity_message(VERBOSE, "Opened image (read only, with privileges) %s\n", source);
+
+            if ( fstat(fileno(image_fp), &image_stat) != 0 ) {
+                singularity_message(ERROR, "Could not obtain stat on image %s: %s\n", source, strerror(errno));
+                ABORT(255);
+            }
+
+            if ( ( file_grp = getgrgid(image_stat.st_gid) ) == NULL ) {
+                singularity_message(ERROR, "Could not obtain gid of image %s: %s\n", source, strerror(errno));
+                ABORT(255);
+            }
+            singularity_priv_drop();
+
+            singularity_config_rewind();
+            do {
+                config_gname = singularity_config_get_value("container group");
+
+                if ( ( has_perm(4, image_stat) == 0 ) || ( has_perm(1, image_stat) == 0 ) ) {
```

Comment:
Can you add a comment here as to what is being tested?  It's not clear what `4` and `1` are... if they are modes, maybe use the equivalent mode macros?